### ISSUE TITLE
false positive from misc-unused-using-decls check with explicit instantiations

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#29740]
|  |  |
| --- | --- |
| Bugzilla Link | [30392](https://llvm.org/bz30392) |
| Version | unspecified |
| OS | Linux |
| Attachments | [Repro file](https://user-images.githubusercontent.com/60944935/143754377-3849d404-3ac6-43bc-ba5d-ec62de3572d4.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @EugeneZelenko,@gribozavr,@hokein |

## Extended Description 

It seems like the 'misc-unused-using-decl' doesn't consider class references in the context of explicit template instantiations. Here is a repro:

```
====== test.cc ======
#include <memory>

namespace N {
  struct Foo {};
} // namespace N

using N::Foo;

template<class T>
void MyTemplateFunction(std::unique_ptr<T>* blah) {
  blah->reset(new T());
}

template
void MyTemplateFunction(std::unique_ptr<Foo>* blah);

=== example output ====
$ ~/sw/clang+llvm-3.9.0-x86_64-linux-gnu-ubuntu-14.04/bin/clang-tidy  --checks=misc-unused-using-decls ./test.cc -- --std=c++11
1 warning generated.
/tmp/test.cc:7:10: warning: using decl 'Foo' is unused [misc-unused-using-decls]
using N::Foo;
         ^
```

Note that removing the 'using' declaration here produces a compile error:

```
/tmp/test.cc:13:41: error: unknown type name 'Foo'; did you mean 'N::Foo'? [clang-diagnostic-error]
void MyTemplateFunction(std::unique_ptr<Foo>* blah);
```

**Comment by carlosgalvezp** [comment]
The example provided by OP and and simpler example afterwards are both fixed in latest trunk.
